src/main/java/org/opensearch/security/http/OnBehalfOfAuthenticator.java

```diff
@@ -0,0 +1,249 @@
+            backendRoles = Arrays.stream(backendRolesObject.toString().split(",")).map(String::trim).toArray(String[]::new);
+        }
+
+        return backendRoles;
+    }
+
+    @Override
+    @SuppressWarnings("removal")
+    public AuthCredentials extractCredentials(RestRequest request, ThreadContext context) throws OpenSearchSecurityException {
+        final SecurityManager sm = System.getSecurityManager();
+
+        if (sm != null) {
+            sm.checkPermission(new SpecialPermission());
+        }
+
+        AuthCredentials creds = AccessController.doPrivileged(new PrivilegedAction<AuthCredentials>() {
+            @Override
+            public AuthCredentials run() {
+                return extractCredentials0(request);
+            }
+        });
+
+        return creds;
+    }
+
+    private AuthCredentials extractCredentials0(final RestRequest request) {
+        if (!oboEnabled) {
+            log.error("On-behalf-of authentication is disabled");
+            return null;
+        }
+
+        if (jwtParser == null) {
+            log.error("Missing Signing Key. JWT authentication will not work");
+            return null;
+        }
+
+        String jwtToken = extractJwtFromHeader(request);
+        if (jwtToken == null) {
+            return null;
+        }
+
+        if (!isAllowedRequest(request)) {
+            return null;
+        }
+
+        try {
+            final Claims claims = jwtParser.parseClaimsJws(jwtToken).getBody();
+
+            final String subject = claims.getSubject();
+            if (subject == null) {
+                log.error("Valid jwt on behalf of token with no subject");
+                return null;
+            }
+
+            final String audience = claims.getAudience();
+            if (audience == null) {
+                log.error("Valid jwt on behalf of token with no audience");
+                return null;
+            }
+
+            final String issuer = claims.getIssuer();
+            if (!issuer.equals(clusterName)) {
+                log.error("The issuer of this OBO does not match the current cluster identifier");
+                return null;
+            }
+
+            List<String> roles = extractSecurityRolesFromClaims(claims);
+            String[] backendRoles = extractBackendRolesFromClaims(claims);
+
+            final AuthCredentials ac = new AuthCredentials(subject, roles, backendRoles).markComplete();
+
+            for (Entry<String, Object> claim : claims.entrySet()) {
+                ac.addAttribute("attr.jwt." + claim.getKey(), String.valueOf(claim.getValue()));
+            }
+
+            return ac;
+
+        } catch (WeakKeyException e) {
+            log.error("Cannot authenticate user with JWT because of ", e);
+        } catch (Exception e) {
+            if (log.isDebugEnabled()) {
+                log.debug("Invalid or expired JWT token.", e);
+            }
+        }
+
+        return null;
```

Comment:
Add a comment, something like 'return null representing that no user was found.'

Reply:
added